### PR TITLE
Small usage changes for air and ignore items

### PR DESCRIPTION
**Pull request by Wuzzy2**
This PR makes some tweaks to air and ignore (in item form):

The “you hacker you!” in the `description` is removed. Instead, this message appears on `/give` and `/giveme`.
Reason: To keep the tooltip clean, and because the `description` might be parsed by mods like `doc` to display item names in a list, so the `description` should be general-purpose.

The placement of `ignore` is prevented straight in `builtin` with a custom error message directed at the player.
Reason: To catch the placement before it goes to the engine, which would cause massive error message spam with cryptic messages.

### CODE REVIEW

File: builtin/game/chatcommands.lua [hunk 1]
```diff
@@ -576,14 +576,27 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
...
 	if giver == receiver then
-		return true, ("%q %sadded to inventory.")
-				:format(stackstring, partiality)
+		if you_hacker then
+			msg = "%q %sadded to inventory (you hacker, you!)."
+		else
+			msg = "%q %sadded to inventory."
+		end
+		msg = msg:format(stackstring, partiality)
+		return true, msg
```
Comment (SmallJoker) [inline builtin/game/chatcommands.lua]: `msg:format(stackstring, partiality)` could be returned directly.

File: builtin/game/chatcommands.lua [hunk 2]
```diff
@@ -576,14 +576,27 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
...
 				:format(stackstring, partiality))
-		return true, ("%q %sadded to %s's inventory.")
-				:format(stackstring, partiality, receiver)
+		if you_hacker then
+			msg = "%q %sadded to %s's inventory (you hacker, you!)."
+		else
+			msg = "%q %sadded to %s's inventory."
+		end
+		msg = msg:format(stackstring, partiality, receiver)
+		return true, msg
```
Comment (SmallJoker) [inline builtin/game/chatcommands.lua]: Ditto.

File: builtin/game/register.lua [hunk 3]
```diff
@@ -368,6 +368,11 @@ core.register_node(":ignore", {
 	air_equivalent = true,
 	drop = "",
 	groups = {not_in_creative_inventory=1},
+	-- Prevent placement of ignore (to prevent engine error message spam in logs)
+	on_place = function(itemstack, placer, pointed_thing)
+		minetest.chat_send_player(placer:get_player_name(), minetest.colorize("#FF0000", "One does not simply place ignore nodes in the world!"))
```
Comment (SmallJoker) [inline builtin/game/register.lua]: Whereas I like the meme potential the line is too long and this message is superfluous. You could return an empty stack instead.
Also: `on_place = function() end` would do the same, but without emptying the stack.

Comment (paramat) [reply builtin/game/register.lua]: I agree, the message just replaces spam with spam.

File: builtin/game/chatcommands.lua [hunk 4]
```diff
@@ -576,14 +576,27 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 	-- The actual item stack string may be different from what the "giver"
 	-- entered (e.g. big numbers are always interpreted as 2^16-1).
 	stackstring = itemstack:to_string()
+	-- easter egg when giving "forbidden" items
+	local you_hacker = itemstack:get_name() == "air" or itemstack:get_name() == "ignore"
```
Comment (paramat) [inline builtin/game/chatcommands.lua]: Perhaps we should actually prevent the giving of 'ignore'? After all it is not placable by a player.

**Comment by paramat** [comment]
I agree with removing that text from the tooltip, but i don't think it needs to be preserved and i feel it doesn't justify the code needed to handle it. It was only worth including as a tooltip because doing so is simple and requires no code.
Instead perhaps we should just prevent the giving of "air" and "ignore"?
:-1: for now.

**Comment by Wuzzy2** [comment]
This was just a quick and dirty idea of mine. I was not sure about it myselves, but I didn't want to throw away my code. I was not sure about the response myselves. :-)

The general idea is just adding some sanity checks in case something goes wrong. A bug in a mod might still somehow spawn `air` or even `ignore` as items; I once had a mod which actually dropped `air`. Of course, these situations should not happen, but when they do, Minetest should fail gracefully. That's the idea behind all this.

I stick to keeping `on_place` because for the above reason the item might still be obtained. But destroying the item on place is a good idea, this would also fix the chat spam.

Giving `air` is not really useful except for teh lulz, but no harm is being done for allowing it.

To summarize the change proposals (with my opinion):

* Keep the shorter `description`s of `air` and `ignore` items
* Prevent giving of `ignore` (“One does not simply give 'ignore' nodes!” :grin: )
* Keep `on_place` of ignore as sanity check (better safe than sorry).
* Destroy `ignore` item on placement. Heck, maybe even the entire item stack, so the message is printed only once

Is this OK?

**Comment by paramat** [comment]
Yes fine, i was somewhat wrong earlier, "ignore" can of course be obtained in other ways than give.

**Comment by paramat** [comment]
It can't be placed so i can't see any use in having it in inventory. I think go ahead, someone can always complain later.

File: builtin/game/chatcommands.lua [hunk 5]
```diff
@@ -559,6 +559,14 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 		return false, "Cannot give an empty item"
 	elseif not itemstack:is_known() then
 		return false, "Cannot give an unknown item"
+	--[[ We forbid giving 'ignore' because it's useless as an item and
+	also a bit dangerous.
+	Note this does NOT make it completely impossible for players to
+	get 'ignore'; they might still obtain it via mods or bugs in mods,
+	so beware. ]]
```
Comment (SmallJoker) [inline builtin/game/chatcommands.lua]: KISS:
`-- Forbid obtaining ignore due to unwanted side effects`

Comment below does not add anything meaningful to the code, please remove it.

File: builtin/game/chatcommands.lua [hunk 6]
```diff
@@ -576,14 +584,26 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 	-- The actual item stack string may be different from what the "giver"
 	-- entered (e.g. big numbers are always interpreted as 2^16-1).
 	stackstring = itemstack:to_string()
+	-- Air is not that useful as an item, but it also does no real harm.
+	-- Thus we do not forbid giving it, but we DO have a snarky comment. ;-)
+	local you_hacker = itemstack:get_name() == "air"
```
Comment (SmallJoker) [inline builtin/game/chatcommands.lua]: Air and other airlike nodes may be given to players in different realms to breathe they don't really cause problems - IMO this check is not required.

File: builtin/game/register.lua [hunk 7]
```diff
@@ -368,6 +368,12 @@ core.register_node(":ignore", {
 	air_equivalent = true,
 	drop = "",
 	groups = {not_in_creative_inventory=1},
+	-- Prevent placement of ignore (to prevent engine error message spam in logs)
+	-- and destroy itemstack.
+	on_place = function(itemstack, placer, pointed_thing)
+		minetest.chat_send_player(placer:get_player_name(), minetest.colorize("#FF0000", "One does not simply place 'ignore' nodes!"))
```
Comment (SmallJoker) [inline builtin/game/register.lua]: Line still needs a break. It's way over the 80 allowed characters.

File: builtin/game/chatcommands.lua (hunk 5 above)
Comment (Wuzzy2) [reply builtin/game/chatcommands.lua]: Huh? The fact that this check does not perfeclty prevent obtaining `ignore` IS absolutely significant, the discussion above shows that paramat did not consider this and would have been tempted to remove neccessary code.

Also: Are you complaining about me making too many code comments? OMG.

```
* Wuzzy dies a little bit inside.
```

File: builtin/game/chatcommands.lua (hunk 6 above)
Comment (Wuzzy2) [reply builtin/game/chatcommands.lua]: You have no sense of humour. :-(
Anyway, has this been actually used/abused on servers? Just curious. :D

File: builtin/game/register.lua (hunk 7 above)
Comment (Wuzzy2) [reply builtin/game/register.lua]: OK.

File: builtin/game/chatcommands.lua (hunk 5 above)
Comment (SmallJoker) [reply builtin/game/chatcommands.lua]: It is obvious that a change in the `/give` command does not affect other mods which can give any item to the player inventory. If there is a place to document the consequences of placing ignore, then it's the wiki page and/or in a Minetest manual. Comments exist to provide additional information which is relevant in this context. For example `so be aware` sounds like a guide entry for "How to make a Minetest mod" - having it here as comment in the code does not help anyone.

File: builtin/game/chatcommands.lua (hunk 6 above)
Comment (SmallJoker) [reply builtin/game/chatcommands.lua]: I'm trying to keep the quality of the code. Rare Easter eggs may be fine but Chuck Norris jokes or memes certainly belong somewhere else.
When I did some testing with the spacesuit and the moon realm generator, I also tried to cut a base into a hill and filling it with regular air to breathe - worked. Air has a low potential for (ab)use, which is probably also why nothing about is heard.

Comment (Wuzzy2) [reply builtin/game/chatcommands.lua]: Note that in the case of `air`, I merely adapted an old easter egg, I did not invent it.
I will remove it if enough core devs believe it must be wiped out.

File: builtin/game/chatcommands.lua (hunk 5 above)
Comment (Wuzzy2) [reply builtin/game/chatcommands.lua]: I repeat again: Paramat (!!!) did not see this initially, so if even core devs can be fooled, I think it is NOT obvious to everyone and a comment is justified. Note I have shortened the comment a bit now, I don't know why it is not visible here.
I like to hear an opinion of another core dev here.

**Comment by Wuzzy2** [comment]
More comment/msg changes …

File: builtin/game/chatcommands.lua (hunk 5 above)
Comment (paramat) [reply builtin/game/chatcommands.lua]: I agree with SmallJoker.

File: builtin/game/chatcommands.lua (hunk 6 above)
Comment (paramat) [reply builtin/game/chatcommands.lua]: I prefer removing the 'you hacker' check and 'you hacker' messages, as i stated above.

File: builtin/game/chatcommands.lua [hunk 8]
```diff
@@ -559,6 +559,12 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 		return false, "Cannot give an empty item"
 	elseif not itemstack:is_known() then
 		return false, "Cannot give an unknown item"
+	--[[ Forbid obtaining 'ignore' due to unwanted side effects.
+	Note this check does NOT make it impossible for players to get 'ignore';
+	they can still obtain it via mods or bugs in mod. ]]
+	elseif itemstack:get_name() == "ignore" then
+		-- A little humour is good, this is a game, after all. :-)
```
[the remark this answers is not in the document]
Comment (paramat) [reply builtin/game/chatcommands.lua]: Agreed.

File: builtin/game/chatcommands.lua [hunk 9]
```diff
@@ -559,6 +559,12 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
...
+		return false, "One does not simpliy give 'ignore' items!"
```
[the remark this answers is not in the document]
Comment (paramat) [reply builtin/game/chatcommands.lua]: Ugh memes.

File: builtin/game/chatcommands.lua [hunk 10]
```diff
@@ -576,14 +579,25 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 	-- The actual item stack string may be different from what the "giver"
 	-- entered (e.g. big numbers are always interpreted as 2^16-1).
 	stackstring = itemstack:to_string()
+	-- Air is not that useful as an item, but it also does no real harm.
+	local you_hacker = itemstack:get_name() == "air"
```
Comment (paramat) [inline builtin/game/chatcommands.lua]: Please remove all 'you hacker' stuff as it's unnecessary code bloat. Then i'll be +1.

File: builtin/game/chatcommands.lua [hunk 11]
```diff
@@ -576,14 +579,25 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 	-- The actual item stack string may be different from what the "giver"
 	-- entered (e.g. big numbers are always interpreted as 2^16-1).
 	stackstring = itemstack:to_string()
+	-- Air is not that useful as an item, but it also does no real harm.
```
Comment (paramat) [inline builtin/game/chatcommands.lua]: No need for this comment.

**Comment by paramat** [comment]
Ok for me once 'you hacker' stuff and unnecessary comment is removed.

**Comment by Wuzzy2** [comment]
@paramat: Done.

I also have sneaked in another commit. I noticed you could give yourself `unknown` as an item by typing in `/giveme unknown`. This caused a couple of weird effects, so I added another check.

**Comment by paramat** [comment]
Thanks :+1: but haven't tested.

File: builtin/game/chatcommands.lua [hunk 12]
```diff
@@ -557,8 +557,11 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
 	local itemstack = ItemStack(stackstring)
 	if itemstack:is_empty() then
 		return false, "Cannot give an empty item"
-	elseif not itemstack:is_known() then
+	elseif (not itemstack:is_known()) or (itemstack:get_name() == "unknown") then
```
Comment (paramat) [inline builtin/game/chatcommands.lua]: Minor point, second set of brackets is not needed.
